buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java

```diff
@@ -83,21 +386,122 @@ private void checkFrozen() {
         }
     }
 
+    private static String safeName(String name) {
+        return name
+            .replaceAll("^[^a-zA-Z0-9]+", "")
+            .replaceAll("[^a-zA-Z0-9]+", "-");
+    }
+
+    private File getHttpPortsFile() {
+        return new File(getConfPathLogs(), "http.ports");
+    }
+
+    private File getTransportPortFile() {
+        return new File(getConfPathLogs(), "transport.ports");
+    }
+
+    private List<String> getTransportPortInternal() {
+        File transportPortFile = getTransportPortFile();
+        try {
+            return readPortsFile(getTransportPortFile());
+        } catch (IOException e) {
+            throw new TestClustersException(
+                "Failed to read transport ports file: " + transportPortFile + " for " + this, e
+            );
+        }
+    }
+
+    private List<String> getHttpPortInternal() {
+        File httpPortsFile = getHttpPortsFile();
+        try {
+            return readPortsFile(getHttpPortsFile());
+        } catch (IOException e) {
+            throw new TestClustersException(
+                "Failed to read http ports file: " + httpPortsFile + " for " + this, e
+            );
+        }
+    }
+
+    private List<String> readPortsFile(File file) throws IOException {
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            return reader.lines()
+                .map(String::trim)
+                .collect(Collectors.toList());
+        }
+    }
+
+    private void waitForAllConditions() {
+        requireNonNull(esProcess, "Can't wait for `" + this + "` as it was stopped.");
+        long startedAt = System.currentTimeMillis();
+        logger.info("Starting to wait for cluster to come up");
+        waitConditions.forEach((description, predicate) -> {
+            long thisConditionStartedAt = System.currentTimeMillis();
+            boolean conditionMet = false;
+            Throwable lastException = null;
+            while (
+                System.currentTimeMillis() - startedAt < MILLISECONDS.convert(NODE_UP_TIMEOUT, NODE_UP_TIMEOUT_UNIT)
+            ) {
+                if (esProcess.isAlive() == false) {
+                    throw new TestClustersException(
+                        "process was found dead while waiting for " + description + ", " + this
+                    );
+                }
+                try {
+                    if(predicate.test(this)) {
+                        conditionMet = true;
+                        break;
+                    }
+                } catch (Exception e) {
+                    if (lastException == null) {
+                        lastException = e;
+                    } else {
+                        e.addSuppressed(lastException);
+                        lastException = e;
+                    }
+                }
+                try {
+                    Thread.sleep(500);
+                }
+                catch (TestClustersException e) {
```

Comment:
We want to keep trying the predicate if an exception occurs, except for `TestClustersException` which we want to bubble up directly. 
It's a way of saying something went wrong while waiting, there's no way it will recover and we want to fail fast. I'll fix the formatting thanks!

Reply:
Ah wait :) doesn't that `try`- `catch` only cover `Thread.sleep` here anyway which will never throw `TestClustersException`?

Reply:
Indeed, it looks like this only covers the sleep.

Reply:
Fixed it, thanks !